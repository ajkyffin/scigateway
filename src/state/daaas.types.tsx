import { Step } from 'react-joyride';

export const AuthSuccessType = 'daaas:auth_success';
export const AuthFailureType = 'daaas:auth_failure';
export const LoadingAuthType = 'daaas:loading_auth';
export const LoadAuthProviderType = 'daaas:loading_auth_provider';
export const TokenExpiredType = 'daaas:expired_token';
export const ConfigureStringsType = 'daaas:configure_strings';
export const LoginType = 'daaas:login';
export const NotificationType = 'daaas:api:notification';
export const RegisterRouteType = 'daaas:api:register_route';
export const RequestPluginRerenderType = 'daaas:api:plugin_rerender';
export const SignOutType = 'daaas:signout';
export const ToggleDrawerType = 'daaas:toggledrawer';
export const DismissNotificationType = 'daaas:dismissnotification';
export const ConfigureFeatureSwitchesType = 'daaas:feature_switches';
export const AddNotificationsType = 'daaas:add_notification';
export const SiteLoadingType = 'daaas:site_loading';
export const ConfigureAnalyticsType = 'daaas:configure_analytics';
export const InitialiseAnalyticsType = 'daaas:initialise_analytics';
export const ToggleHelpType = 'daaas:toggle_help';
export const AddHelpTourStepsType = 'daaas:add_help_tour_steps';

export interface NotificationPayload {
  message: string;
  severity: string;
}

export interface AddNotificationsPayload {
  notifications: NotificationPayload[];
}

export interface AppStrings {
  [id: string]: string;
}

export interface ConfigureStringsPayload {
  res: ApplicationStrings;
}

export interface ApplicationStrings {
  [section: string]: AppStrings;
}

export interface FeatureSwitchesPayload {
  switches: FeatureSwitches;
}

export interface FeatureSwitches {
  showContactButton: boolean;
}

export interface RegisterRoutePayload {
  section: string;
  link: string;
  plugin: string;
  displayName: string;
  order: number;
  helpText?: string;
}

export interface PluginConfig {
  section: string;
  link: string;
  plugin: string;
  displayName: string;
  order: number;
  helpText?: string;
}

export interface GroupedPlugins {
  [section: string]: PluginConfig[];
}

export interface LoginPayload {
  username: string;
  password: string;
}

export interface AccessPayload {
  authorisation: boolean;
  accessToken: string;
  error: null;
}

export interface AuthorisedPayload {
  token: string;
}

export interface DismissNotificationPayload {
  index: number;
}

export interface AuthProviderPayload {
  authProvider: string;
}

export interface SiteLoadingPayload {
  loading: boolean;
}

export interface ConfigureAnalyticsPayload {
  id: string;
}

export interface AddHelpTourStepsPayload {
  steps: Step[];
}
