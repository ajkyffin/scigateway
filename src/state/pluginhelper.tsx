import {
  GroupedPlugins,
  PluginConfig,
  RegisterRoutePayload,
} from './daaas.types';

// Custom comparer for plugin objects, sorts keys by order (asc), then by name (asc)
export const comparePlugins = (a: PluginConfig, b: PluginConfig): number =>
  a.order === b.order
    ? a.displayName.localeCompare(b.displayName)
    : a.order - b.order;

function buildMenuItemDictionary(plugins: PluginConfig[]): GroupedPlugins {
  const dict: GroupedPlugins = {};

  plugins.forEach(p => {
    if (!(p.section in dict)) {
      dict[p.section] = [];
    }
    dict[p.section].push(p);
  });

  return dict;
}

// Convert the list of plugins into a structured dataset for rendering
export function structureMenuData(plugins: PluginConfig[]): GroupedPlugins {
  const menuItems = buildMenuItemDictionary(plugins);

  Object.keys(menuItems).forEach(
    section => (menuItems[section] = menuItems[section].sort(comparePlugins))
  );
  return menuItems;
}

// Convert a plugin registration request body into a pluginConfig object
export function buildPluginConfig(payload: RegisterRoutePayload): PluginConfig {
  return { ...payload };
}
