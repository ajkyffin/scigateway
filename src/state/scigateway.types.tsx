import { Step } from 'react-joyride';

export const AuthSuccessType = 'scigateway:auth_success';
export const AuthFailureType = 'scigateway:auth_failure';
export const LoadingAuthType = 'scigateway:loading_auth';
export const LoadAuthProviderType = 'scigateway:loading_auth_provider';
export const LoadAuthUrlType = 'scigateway:loading_auth_url';
export const InvalidateTokenType = 'scigateway:api:invalidate_token';
export const ConfigureStringsType = 'scigateway:configure_strings';
export const LoginType = 'scigateway:login';
export const NotificationType = 'scigateway:api:notification';
export const RegisterRouteType = 'scigateway:api:register_route';
export const RequestPluginRerenderType = 'scigateway:api:plugin_rerender';
export const SignOutType = 'scigateway:signout';
export const ToggleDrawerType = 'scigateway:toggledrawer';
export const DismissNotificationType = 'scigateway:dismissnotification';
export const ConfigureFeatureSwitchesType = 'scigateway:feature_switches';
export const AddNotificationsType = 'scigateway:add_notification';
export const SiteLoadingType = 'scigateway:site_loading';
export const ConfigureAnalyticsType = 'scigateway:configure_analytics';
export const InitialiseAnalyticsType = 'scigateway:initialise_analytics';
export const ToggleHelpType = 'scigateway:toggle_help';
export const AddHelpTourStepsType = 'scigateway:add_help_tour_steps';

export interface NotificationPayload {
  message: string;
  severity: string;
}

export interface AddNotificationsPayload {
  notifications: NotificationPayload[];
}

export interface AppStrings {
  [id: string]: string;
}

export interface ConfigureStringsPayload {
  res: ApplicationStrings;
}

export interface ApplicationStrings {
  [section: string]: AppStrings;
}

export interface FeatureSwitchesPayload {
  switches: FeatureSwitches;
}

export interface FeatureSwitches {
  showContactButton: boolean;
}

export interface RegisterRoutePayload {
  section: string;
  link: string;
  plugin: string;
  displayName: string;
  order: number;
  helpText?: string;
}

export interface PluginConfig {
  section: string;
  link: string;
  plugin: string;
  displayName: string;
  order: number;
  helpText?: string;
}

export interface GroupedPlugins {
  [section: string]: PluginConfig[];
}

export interface LoginPayload {
  username: string;
  password: string;
}

export interface AccessPayload {
  authorisation: boolean;
  accessToken: string;
  error: null;
}

export interface AuthorisedPayload {
  token: string;
}

export interface DismissNotificationPayload {
  index: number;
}

export interface AuthProviderPayload {
  authProvider: string;
  authUrl: string | undefined;
}

export interface SiteLoadingPayload {
  loading: boolean;
}

export interface ConfigureAnalyticsPayload {
  id: string;
}

export interface AddHelpTourStepsPayload {
  steps: Step[];
}
