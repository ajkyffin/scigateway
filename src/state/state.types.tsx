import { ThunkAction } from 'redux-thunk';
import { AnyAction } from 'redux';
import { Step } from 'react-joyride';
import {
  ApplicationStrings,
  PluginConfig,
  FeatureSwitches,
} from './scigateway.types';
import { RouterState } from 'connected-react-router';

export interface Plugin {
  name: string;
  src: string;
  enable: boolean;
  location: 'main' | 'left' | 'right';
}

export interface ScigatewayNotification {
  message: string;
  severity: string;
}

export interface ScigatewayState {
  notifications: ScigatewayNotification[];
  plugins: PluginConfig[];
  drawerOpen: boolean;
  siteLoading: boolean;
  showHelp: boolean;
  helpSteps: Step[];
  authorisation: AuthState;
  res?: ApplicationStrings;
  features: FeatureSwitches;
  analytics?: AnalyticsState;
}

export interface StateType {
  scigateway: ScigatewayState;
  router: RouterState;
}

export interface ActionType<T> {
  type: string;
  payload: T;
}

export type ThunkResult<R> = ThunkAction<R, StateType, null, AnyAction>;

export interface User {
  username: string;
  avatarUrl: string;
}

// eslint-disable-next-line @typescript-eslint/interface-name-prefix
export interface ICATAuthenticator {
  mnemonic: string;
  keys: { name: string; hide?: boolean }[];
  friendly?: string;
  admin?: boolean;
}

export interface AuthProvider {
  isLoggedIn: () => boolean;
  logOut: () => void;
  logIn: (username: string, password: string) => Promise<void>;
  verifyLogIn: () => Promise<void>;
  refresh: () => Promise<void>;
  redirectUrl: string | null;
  user: User | null;
  mnemonic?: string;
  autoLogin?: () => Promise<void>;
}

export interface AuthState {
  failedToLogin: boolean;
  signedOutDueToTokenInvalidation: boolean;
  loading: boolean;
  provider: AuthProvider;
}

export interface AnalyticsState {
  id: string;
  initialised: boolean;
}
