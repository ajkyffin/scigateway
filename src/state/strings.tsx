import { AppStrings } from './scigateway.types';
import { StateType } from './state.types';

export function getAppStrings(
  state: StateType,
  section: string
): AppStrings | undefined {
  return state.scigateway.res ? state.scigateway.res[section] : undefined;
}

export const getString = (res: AppStrings | undefined, key: string): string =>
  (res && res[key]) || key;
