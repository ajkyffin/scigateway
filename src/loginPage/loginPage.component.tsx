import React, { useState, useEffect } from 'react';
import { connect } from 'react-redux';
import { AnyAction } from 'redux';
import { ThunkDispatch } from 'redux-thunk';
import Avatar from '@material-ui/core/Avatar';
import LockOutlinedIcon from '@material-ui/icons/LockOutlined';
import TextField from '@material-ui/core/TextField';
import Button from '@material-ui/core/Button';
import Paper from '@material-ui/core/Paper';
import CircularProgress from '@material-ui/core/CircularProgress';
import Typography from '@material-ui/core/Typography';
import {
  withStyles,
  Theme,
  StyleRules,
  createStyles,
  WithStyles,
} from '@material-ui/core/styles';
import { verifyUsernameAndPassword } from '../state/actions/scigateway.actions';
import { AppStrings } from '../state/scigateway.types';
import { StateType, AuthState } from '../state/state.types';
import { UKRITheme } from '../theming';
import { getAppStrings, getString } from '../state/strings';
import { Location } from 'history';

const styles = (theme: Theme): StyleRules =>
  createStyles({
    root: {
      display: 'flex',
      flexDirection: 'column',
      alignItems: 'center',
      width: 'auto',
      marginLeft: theme.spacing(3),
      marginRight: theme.spacing(3),
    },
    avatar: {
      // material-ui v4 upgrade
      margin: theme.spacing(1),
      backgroundColor: (theme as UKRITheme).ukri.orange,
    },
    paper: {
      marginTop: theme.spacing(8),
      display: 'flex',
      flexDirection: 'column',
      alignItems: 'center',
      // material-ui v4 upgrade
      padding: `${theme.spacing(3)}px ${theme.spacing(3)}px ${theme.spacing(
        3
      )}px`,
      [theme.breakpoints.up(400 + theme.spacing(6))]: {
        width: 400,
        marginLeft: 'auto',
        marginRight: 'auto',
      },
    },
    textField: {
      // material-ui v4 upgrade
      marginTop: theme.spacing(1),
    },
    button: {
      // material-ui v4 upgrade
      marginTop: `${theme.spacing(1)}px`,
    },
    warning: {
      // material-ui v4 upgrade
      marginTop: `${theme.spacing(1)}px`,
      color: 'red',
    },
    info: {
      // material-ui v4 upgrade
      marginTop: `${theme.spacing(1)}px`,
      color: theme.palette.primary.main,
    },
    spinner: {
      marginTop: 15,
    },
  });

interface LoginPageProps {
  auth: AuthState;
  res: AppStrings | undefined;
  location: Location;
}

interface LoginPageDispatchProps {
  verifyUsernameAndPassword: (
    username: string,
    password: string
  ) => Promise<void>;
}

export type CombinedLoginProps = LoginPageProps &
  LoginPageDispatchProps &
  WithStyles<typeof styles>;

export const RedirectLoginScreen = (
  props: CombinedLoginProps
): React.ReactElement => (
  <div className={props.classes.root}>
    {props.auth.failedToLogin ? (
      <Typography className={props.classes.warning}>
        {getString(props.res, 'login-redirect-error-msg')}
      </Typography>
    ) : null}
    <Button
      variant="contained"
      color="primary"
      className={props.classes.button}
      disabled={props.auth.loading}
      onClick={() => {
        if (props.auth.provider.redirectUrl) {
          window.location.href = props.auth.provider.redirectUrl;
        }
      }}
    >
      <Typography color="inherit" noWrap style={{ marginTop: 3 }}>
        Login with Github
      </Typography>
    </Button>
  </div>
);

export const CredentialsLoginScreen = (
  props: CombinedLoginProps
): React.ReactElement => {
  const [username, setUsername] = useState<string>('');
  const [password, setPassword] = useState<string>('');

  const isInputValid = (): boolean => username !== '' && password !== '';

  return (
    <div
      className={props.classes.root}
      onKeyPress={e => {
        if (
          !props.auth.provider.redirectUrl &&
          e.key === 'Enter' &&
          isInputValid()
        ) {
          props.verifyUsernameAndPassword(username, password);
        }
      }}
    >
      {props.auth.failedToLogin ? (
        <Typography className={props.classes.warning}>
          {getString(props.res, 'login-error-msg')}
        </Typography>
      ) : null}
      {props.auth.signedOutDueToTokenInvalidation ? (
        <Typography className={props.classes.info}>
          {getString(props.res, 'token-invalid-msg')}
        </Typography>
      ) : null}
      <TextField
        className={props.classes.textField}
        label={getString(props.res, 'username-placeholder')}
        value={username}
        onChange={e => setUsername(e.currentTarget.value)}
        disabled={props.auth.loading}
      />
      <TextField
        className={props.classes.textField}
        label={getString(props.res, 'password-placeholder')}
        value={password}
        onChange={e => setPassword(e.currentTarget.value)}
        type="password"
        disabled={props.auth.loading}
      />
      <Button
        variant="contained"
        color="primary"
        className={props.classes.button}
        disabled={!isInputValid() || props.auth.loading}
        onClick={() => {
          props.verifyUsernameAndPassword(username, password);
        }}
      >
        <Typography color="inherit" noWrap style={{ marginTop: 3 }}>
          {getString(props.res, 'login-button')}
        </Typography>
      </Button>
    </div>
  );
};

const LoginPageComponent = (props: CombinedLoginProps): React.ReactElement => {
  useEffect(() => {
    if (
      props.auth.provider.redirectUrl &&
      props.location.search &&
      !props.auth.loading &&
      !props.auth.failedToLogin
    ) {
      if (props.location.search) {
        props.verifyUsernameAndPassword('', props.location.search);
      }
    }
  });

  return (
    <div className={props.classes.root}>
      <Paper className={props.classes.paper}>
        <Avatar className={props.classes.avatar}>
          <LockOutlinedIcon />
        </Avatar>
        <Typography component="h1" variant="h5">
          {getString(props.res, 'title')}
        </Typography>
        {props.auth.provider.redirectUrl ? (
          <RedirectLoginScreen {...props} />
        ) : (
          <CredentialsLoginScreen {...props} />
        )}
        {props.auth.loading ? (
          <CircularProgress className={props.classes.spinner} />
        ) : null}
      </Paper>
    </div>
  );
};

const mapStateToProps = (state: StateType): LoginPageProps => ({
  auth: state.scigateway.authorisation,
  res: getAppStrings(state, 'login'),
  location: state.router.location,
});

const mapDispatchToProps = (
  dispatch: ThunkDispatch<StateType, null, AnyAction>
): LoginPageDispatchProps => ({
  verifyUsernameAndPassword: (username, password) =>
    dispatch(verifyUsernameAndPassword(username, password)),
});

export const LoginPageWithoutStyles = LoginPageComponent;
export const LoginPageWithStyles = withStyles(styles)(LoginPageComponent);

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(LoginPageWithStyles);
