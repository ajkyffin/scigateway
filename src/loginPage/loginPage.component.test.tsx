import React from 'react';
import { LoginPageWithStyles } from './loginPage.component';
import { createShallow, createMount } from '@material-ui/core/test-utils';
import { StateType } from '../state/state.types';
import configureStore from 'redux-mock-store';
import { initialState } from '../state/reducers/daaas.reducer';
import { RouterState } from 'connected-react-router';

describe('Login page component', () => {
  let shallow;
  let mount;
  let mockStore;
  let state: StateType;

  const routerState: RouterState = {
    action: 'POP',
    location: {
      hash: '',
      key: '',
      pathname: '/',
      search: '',
      state: {},
    },
  };

  beforeEach(() => {
    shallow = createShallow({});
    mount = createMount();

    mockStore = configureStore();

    state = {
      daaas: initialState,
      router: routerState,
    };
  });

  it('login page renders correctly', () => {
    const wrapper = shallow(
      <LoginPageWithStyles
        store={mockStore(state)}
        auth={state.daaas.authorisation}
      />
    );
    expect(wrapper.dive()).toMatchSnapshot();
  });

  it('on submit verification method should be called with username and password arguments', async () => {
    const mockLoginfn = jest.fn();
    const testStore = mockStore(state);

    const wrapper = mount(
      <LoginPageWithStyles
        verifyUsernameAndPassword={mockLoginfn}
        auth={state.daaas.authorisation}
        store={testStore}
      />
    );

    const simulateUsernameInput = wrapper.find('input').at(0);
    simulateUsernameInput.instance().value = 'new username';
    simulateUsernameInput.simulate('change');

    const simulatePasswordInput = wrapper.find('input').at(1);
    simulatePasswordInput.instance().value = 'new password';
    simulatePasswordInput.simulate('change');

    wrapper.find('button').simulate('click');

    expect(mockLoginfn.mock.calls.length).toEqual(1);

    expect(mockLoginfn.mock.calls[0]).toEqual(['new username', 'new password']);
  });
});
