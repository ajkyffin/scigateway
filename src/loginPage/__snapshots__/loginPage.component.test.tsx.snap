// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login page component credential component renders correctly 1`] = `
<div
  className="root-1"
  onKeyPress={[Function]}
>
  <TextField
    disabled={false}
    label="username-placeholder"
    onChange={[Function]}
    required={false}
    select={false}
    value=""
    variant="standard"
  />
  <TextField
    disabled={false}
    label="password-placeholder"
    onChange={[Function]}
    required={false}
    select={false}
    type="password"
    value=""
    variant="standard"
  />
  <WithStyles(Button)
    color="primary"
    disabled={true}
    onClick={[Function]}
    variant="contained"
  >
    <WithStyles(Typography)
      color="inherit"
      noWrap={true}
      style={
        Object {
          "marginTop": 3,
        }
      }
    >
      login-button
    </WithStyles(Typography)>
  </WithStyles(Button)>
</div>
`;

exports[`Login page component login page renders credential component if no redirect url 1`] = `
<div
  className="LoginPageComponent-root-1 root-1"
>
  <WithStyles(Paper)
    className="LoginPageComponent-paper-3"
  >
    <WithStyles(Avatar)
      className="LoginPageComponent-avatar-2"
    >
      <pure(LockOutlinedIcon) />
    </WithStyles(Avatar)>
    <WithStyles(Typography)
      component="h1"
      variant="h5"
    >
      title
    </WithStyles(Typography)>
    <CredentialsLoginScreen
      auth={
        Object {
          "failedToLogin": false,
          "loading": false,
          "provider": TestAuthProvider {
            "redirectUrl": null,
            "token": null,
            "user": null,
          },
          "signedOutDueToTokenExpiry": false,
        }
      }
      classes={
        Object {
          "avatar": "LoginPageComponent-avatar-2",
          "button": "LoginPageComponent-button-5",
          "info": "LoginPageComponent-info-7",
          "paper": "LoginPageComponent-paper-3",
          "root": "LoginPageComponent-root-1 root-1",
          "spinner": "LoginPageComponent-spinner-8",
          "textField": "LoginPageComponent-textField-4",
          "warning": "LoginPageComponent-warning-6",
        }
      }
      location={
        Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        }
      }
      verifyUsernameAndPassword={[Function]}
    />
  </WithStyles(Paper)>
</div>
`;

exports[`Login page component login page renders redirect component if redirect url present 1`] = `
<div
  className="LoginPageComponent-root-1 root-1"
>
  <WithStyles(Paper)
    className="LoginPageComponent-paper-3"
  >
    <WithStyles(Avatar)
      className="LoginPageComponent-avatar-2"
    >
      <pure(LockOutlinedIcon) />
    </WithStyles(Avatar)>
    <WithStyles(Typography)
      component="h1"
      variant="h5"
    >
      title
    </WithStyles(Typography)>
    <RedirectLoginScreen
      auth={
        Object {
          "failedToLogin": false,
          "loading": false,
          "provider": TestAuthProvider {
            "redirectUrl": "test redirect",
            "token": null,
            "user": null,
          },
          "signedOutDueToTokenExpiry": false,
        }
      }
      classes={
        Object {
          "avatar": "LoginPageComponent-avatar-2",
          "button": "LoginPageComponent-button-5",
          "info": "LoginPageComponent-info-7",
          "paper": "LoginPageComponent-paper-3",
          "root": "LoginPageComponent-root-1 root-1",
          "spinner": "LoginPageComponent-spinner-8",
          "textField": "LoginPageComponent-textField-4",
          "warning": "LoginPageComponent-warning-6",
        }
      }
      location={
        Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        }
      }
      verifyUsernameAndPassword={[Function]}
    />
  </WithStyles(Paper)>
</div>
`;

exports[`Login page component redirect component renders correctly 1`] = `
<div
  className="root-1"
>
  <WithStyles(Button)
    color="primary"
    disabled={false}
    onClick={[Function]}
    variant="contained"
  >
    <WithStyles(Typography)
      color="inherit"
      noWrap={true}
      style={
        Object {
          "marginTop": 3,
        }
      }
    >
      Login with Github
    </WithStyles(Typography)>
  </WithStyles(Button)>
</div>
`;
